fpga_interchange/examples/chipdb.cmake

```diff
@@ -0,0 +1,290 @@
+    set(oneValueArgs device part)
+    set(multiValueArgs)
+
+    cmake_parse_arguments(
+        create_rapidwright_device_db
+        "${options}"
+        "${oneValueArgs}"
+        "${multiValueArgs}"
+        ${ARGN}
+    )
+
+    set(device ${create_rapidwright_device_db_device})
+    set(part ${create_rapidwright_device_db_part})
+
+    create_rapidwright_device_db(
+        device ${device}
+        part ${part}
+        output_target rapidwright_device
+    )
+
+    # Generate constraints patch
+    create_patched_device_db(
+        device ${device}
+        patch_name constraints
+        patch_path constraints
+        patch_format yaml
+        patch_data ${PYTHON_INTERCHANGE_PATH}/test_data/series7_constraints.yaml
+        input_device ${rapidwright_device}
+        output_target constraints_device
+    )
+
+    # Generate lut constraints patch
+    create_patched_device_db(
+        device ${device}
+        patch_name constraints-luts
+        patch_path lutDefinitions
+        patch_format yaml
+        patch_data ${PYTHON_INTERCHANGE_PATH}/test_data/series7_luts.yaml
+        input_device ${constraints_device}
+        output_target constraints_luts_device
+    )
+
+    set(device_target ${constraints_luts_device} PARENT_SCOPE)
+endfunction()
+
+function(generate_chipdb)
+    # ~~~
+    # generate_chipdb(
+    #    device <common device>
+    #    part <part>
+    #    device_target <device target>
+    #    bel_bucket_seeds <bel bucket seeds>
+    #    package <package>
+    # )
+    # ~~~
+    #
+    # Generates a chipdb BBA file, starting from a device database.
+    #
+    # The chipdb binary file is directly generated to the
+    # <nextpnr-root>/build/fpga_interchange/chipdb/ directory.
+    #
+    # The package argument is only used to run the architecture check target.
+    #
+    # Targets generated:
+    #   - chipdb-<device>-bba
+
+    set(options)
+    set(oneValueArgs device part device_target bel_bucket_seeds package)
+    set(multiValueArgs)
+
+    cmake_parse_arguments(
+        generate_chipdb
+        "${options}"
+        "${oneValueArgs}"
+        "${multiValueArgs}"
+        ${ARGN}
+    )
+
+    set(device ${generate_chipdb_device})
+    set(part ${generate_chipdb_part})
+    set(device_target ${generate_chipdb_device_target})
+    set(bel_bucket_seeds ${generate_chipdb_bel_bucket_seeds})
+    set(package ${generate_chipdb_package})
+
+    get_target_property(device_loc ${device_target} LOCATION)
+    set(chipdb_bba ${CMAKE_CURRENT_BINARY_DIR}/chipdb-${device}.bba)
+    add_custom_command(
+        OUTPUT ${chipdb_bba}
+        COMMAND
+            ${PYTHON_EXECUTABLE} -mfpga_interchange.nextpnr_emit
+                --schema_dir ${INTERCHANGE_SCHEMA_PATH}
+                --output_dir ${CMAKE_CURRENT_BINARY_DIR}
+                --bel_bucket_seeds ${bel_bucket_seeds}
+                --device ${device_loc}
+        COMMAND
+            mv ${CMAKE_CURRENT_BINARY_DIR}/chipdb.bba ${chipdb_bba}
```

Comment:
This is a dangerous solution, because in a parallel build this could result in a file collision.  I recommend we run `nextpnr_emit` in a unique working directory and use it in place, rather than moving it.

Reply:
Done. Each device already had its own directory, so now, the bba will just be called chipdb.bba